File: host/Sources/VibeWireHost/Claude/ClaudeSessionIndex.swift
import Foundation

/// Reads Claude Code's on-disk session history so the phone can list and resume
/// real past conversations — the "connect to my Claude Code chats" requirement.
///
/// Claude Code stores one JSONL transcript per session under
/// `~/.claude/projects/<slugified-cwd>/<session-uuid>.jsonl`. Nothing here
/// writes; resuming is done by handing the session id to the CLI.
struct ClaudeSessionSummary: Sendable {
    let id: String
    let summary: String
    let cwd: String
    let gitBranch: String?
    let modifiedAt: Date
    let messageCount: Int

    var wire: [String: Any] {
        var payload: [String: Any] = [
            "id": id,
            "summary": summary,
            "cwd": cwd,
            "modifiedAt": ISO8601DateFormatter().string(from: modifiedAt),
            "messages": messageCount,
        ]
        payload["gitBranch"] = gitBranch
        return payload
    }
}

enum ClaudeSessionIndex {
    static var projectsDirectory: URL {
        FileManager.default.homeDirectoryForCurrentUser
            .appendingPathComponent(".claude/projects", isDirectory: true)
    }

    /// Claude Code slugifies the working directory by replacing path separators
    /// and dots with dashes. Matching that lets us scope the list to one repo.
    static func slug(for path: String) -> String {
        var slug = ""
        for character in path {
            if character.isLetter || character.isNumber {
                slug.append(character)
            } else {
                slug.append("-")
            }
        }
        return slug
    }

    /// Most recent sessions first. `cwd` scopes to one project; nil lists all.
    static func recentSessions(cwd: String?, limit: Int = 25) -> [ClaudeSessionSummary] {
        let fileManager = FileManager.default
        let root = projectsDirectory

        var projectDirectories: [URL] = []
        if let cwd {
            let directory = root.appendingPathComponent(slug(for: cwd), isDirectory: true)
            if fileManager.fileExists(atPath: directory.path) {
                projectDirectories = [directory]
            }
        } else {
            projectDirectories = (try? fileManager.contentsOfDirectory(
                at: root,
                includingPropertiesForKeys: [.isDirectoryKey],
                options: [.skipsHiddenFiles]
            )) ?? []
        }

        var summaries: [ClaudeSessionSummary] = []
        for directory in projectDirectories {
            let files = (try? fileManager.contentsOfDirectory(
                at: directory,
                includingPropertiesForKeys: [.contentModificationDateKey],
                options: [.skipsHiddenFiles]
            )) ?? []

            for file in files where file.pathExtension == "jsonl" {
                guard let summary = parseTranscript(file) else { continue }
                summaries.append(summary)
            }
        }

        return summaries
            .sorted { $0.modifiedAt > $1.modifiedAt }
            .prefix(limit)
            .map { $0 }
    }

    /// One past message, for repopulating the panel when a session is resumed.
    struct Turn: Sendable {
        let role: String        // "user" | "assistant"
        let text: String
        let at: Date?

        var wire: [String: Any] {
            var payload: [String: Any] = ["role": role, "text": text]
            if let at { payload["at"] = ISO8601DateFormatter().string(from: at) }
            return payload
        }
    }

    /// The conversation so far.
    ///
    /// `claude --resume` under `--print` restores the context for the model but
    /// prints nothing, so a resumed session arrives on the phone as an empty
    /// panel. The transcript on disk is the only copy of what was already said,
    /// and reading it is what makes resume look like resume.
    ///
    /// Returns the last `limit` turns; long messages are clipped because this
    /// goes over the wire to a phone.
    static func transcript(sessionId: String, limit: Int = 40) -> [Turn] {
        guard let url = transcriptURL(for: sessionId),
              let contents = try? String(contentsOf: url, encoding: .utf8)
        else { return [] }

        let formatter = ISO8601DateFormatter()
        var turns: [Turn] = []

        for line in contents.split(separator: "\n") {
            guard let object = try? JSONSerialization.jsonObject(
                with: Data(line.utf8)
            ) as? [String: Any] else { continue }

            let type = object["type"] as? String
            guard type == "user" || type == "assistant" else { continue }
            // Injected context, not something either party said.
            guard (object["isMeta"] as? Bool) != true else { continue }
            guard let message = object["message"] as? [String: Any],
                  let text = extractText(from: message["content"]),
                  !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
            else { continue }

            // Tool plumbing and harness scaffolding are not conversation.
            let lowered = text.lowercased()
            guard !lowered.hasPrefix("<command"),
                  !lowered.hasPrefix("caveat:"),
                  !lowered.hasPrefix("<local-command"),
                  !lowered.contains("</system-reminder>")
            else { continue }

            turns.append(Turn(
                role: type == "user" ? "user" : "assistant",
                text: String(text.prefix(4000)),
                at: (object["timestamp"] as? String).flatMap(formatter.date(from:))
            ))
        }

        return Array(turns.suffix(limit))
    }

    /// Transcripts are filed under a slug of the working directory, and the
    /// caller resuming by id does not necessarily know which one — so search.
    private static func transcriptURL(for sessionId: String) -> URL? {
        let fileManager = FileManager.default
        let directories = (try? fileManager.contentsOfDirectory(
            at: projectsDirectory,
            includingPropertiesForKeys: nil,
            options: [.skipsHiddenFiles]
        )) ?? []

        for directory in directories {
            let candidate = directory.appendingPathComponent("\(sessionId).jsonl")
            if fileManager.fileExists(atPath: candidate.path) { return candidate }
        }
        return nil
    }

    /// Pulls just enough out of a transcript to render a list row: the first
    /// user message as a title, plus cwd and branch from the session metadata.
    ///
    /// Transcripts can be megabytes. Reading the whole file to show a one-line
    /// summary would make the list sluggish, so this reads the head for
    /// metadata and counts lines without materialising the parsed objects.
    private static func parseTranscript(_ url: URL) -> ClaudeSessionSummary? {
        guard let handle = try? FileHandle(forReadingFrom: url) else { return nil }
        defer { try? handle.close() }

        let attributes = try? FileManager.default.attributesOfItem(atPath: url.path)
        let modifiedAt = (attributes?[.modificationDate] as? Date) ?? Date.distantPast

        // 256 KB is plenty for the first several messages.
        let head = (try? handle.read(upToCount: 256 * 1024)) ?? Data()
        guard !head.isEmpty else { return nil }

        var cwd = ""
        var gitBranch: String?
        var title: String?

        let lines = head.split(separator: UInt8(ascii: "\n"))
        for line in lines {
            guard let object = try? JSONSerialization.jsonObject(with: Data(line)) as? [String: Any]
            else { continue }

            if cwd.isEmpty, let value = object["cwd"] as? String { cwd = value }
            if gitBranch == nil, let value = object["gitBranch"] as? String, !value.isEmpty {
                gitBranch = value
            }

            if title == nil,
               (object["type"] as? String) == "user",
               let message = object["message"] as? [String: Any],
               let candidate = extractText(from: message["content"]),
               let cleaned = summaryText(from: candidate) {
                title = cleaned
            }

            if title != nil && !cwd.isEmpty && gitBranch != nil { break }
        }

        // Line count over the whole file, streamed.
        var messageCount = 0
        if let counter = try? FileHandle(forReadingFrom: url) {
            defer { try? counter.close() }
            while let chunk = try? counter.read(upToCount: 1 << 20), !chunk.isEmpty {
                messageCount += chunk.reduce(into: 0) { total, byte in
                    if byte == UInt8(ascii: "\n") { total += 1 }
                }
            }
        }

        let id = url.deletingPathExtension().lastPathComponent
        let summary = (title?.isEmpty == false ? title! : "Untitled session")
            .replacingOccurrences(of: "\n", with: " ")
            .trimmingCharacters(in: .whitespacesAndNewlines)

        return ClaudeSessionSummary(
            id: id,
            summary: String(summary.prefix(120)),
            cwd: cwd.isEmpty ? url.deletingLastPathComponent().lastPathComponent : cwd,
            gitBranch: gitBranch,
            modifiedAt: modifiedAt,
            messageCount: messageCount
        )
    }

    /// Transcripts often open with tooling scaffolding — command wrappers,
    /// caveat banners, injected reminders — none of which describe what the
    /// session was about. Strip those and take the first line that reads
    /// like something a person typed.
    private static func summaryText(from raw: String) -> String? {
        var text = raw

        // Drop any XML-ish wrapper blocks and their contents.
        while let open = text.range(of: "<"),
              let closeTag = text.range(of: ">", range: open.upperBound..<text.endIndex) {
            let tag = String(text[open.upperBound..<closeTag.lowerBound])
                .split(separator: " ").first.map(String.init) ?? ""
            guard !tag.isEmpty, !tag.hasPrefix("/") else { break }
            if let end = text.range(of: "</\(tag)>") {
                text.removeSubrange(open.lowerBound..<end.upperBound)
            } else {
                text.removeSubrange(open.lowerBound..<closeTag.upperBound)
            }
        }

        let ignored = ["caveat:", "system-reminder", "<command", "local-command"]
        for line in text.split(separator: "\n") {
            let trimmed = line.trimmingCharacters(in: .whitespacesAndNewlines)
            guard trimmed.count > 3 else { continue }
            let lowered = trimmed.lowercased()
            guard !ignored.contains(where: lowered.hasPrefix) else { continue }
            return trimmed
        }
        return nil
    }

    /// Message content is either a plain string or an array of typed blocks.
    private static func extractText(from content: Any?) -> String? {
        if let text = content as? String { return text }
        guard let blocks = content as? [[String: Any]] else { return nil }
        for block in blocks where (block["type"] as? String) == "text" {
            if let text = block["text"] as? String, !text.isEmpty { return text }
        }
        return nil
    }
}

File: host-electron/src/main/claude/claudeSessionIndex.ts
import { closeSync, existsSync, openSync, readdirSync, readSync, statSync } from 'node:fs'
import { basename, join } from 'node:path'

/**
 * Reads Claude Code's on-disk session history so the phone can list and resume
 * real past conversations. One JSONL transcript per session under
 * `~/.claude/projects/<slugified-cwd>/<session-uuid>.jsonl`. Nothing here
 * writes; resuming is done by handing the session id to the CLI.
 */
export interface ClaudeSessionSummary {
  id: string
  summary: string
  cwd: string
  gitBranch: string | null
  modifiedAt: Date
  messageCount: number
}

export function sessionWire(session: ClaudeSessionSummary): Record<string, unknown> {
  const payload: Record<string, unknown> = {
    id: session.id,
    summary: session.summary,
    cwd: session.cwd,
    modifiedAt: session.modifiedAt.toISOString(),
    messages: session.messageCount,
  }
  if (session.gitBranch) payload.gitBranch = session.gitBranch
  return payload
}

/** One past message, for repopulating the panel when a session is resumed. */
export interface Turn {
  role: 'user' | 'assistant'
  text: string
  at: Date | null
}

export function turnWire(turn: Turn): Record<string, unknown> {
  const payload: Record<string, unknown> = { role: turn.role, text: turn.text }
  if (turn.at) payload.at = turn.at.toISOString()
  return payload
}

type Json = Record<string, unknown>

export class ClaudeSessionIndex {
  constructor(private readonly projectsDirectory: string) {}

  /** Claude Code slugifies the working directory: every character that is not
   *  a letter or a digit becomes a dash. Matching that scopes the list to one repo. */
  static slug(path: string): string {
    return [...path].map((character) => (/[\p{L}\p{N}]/u.test(character) ? character : '-')).join('')
  }

  /** Most recent sessions first. `cwd` scopes to one project; null lists all. */
  recentSessions(cwd: string | null, limit = 25): ClaudeSessionSummary[] {
    let directories: string[] = []
    if (cwd) {
      const directory = join(this.projectsDirectory, ClaudeSessionIndex.slug(cwd))
      if (existsSync(directory)) directories = [directory]
    } else {
      directories = this.projectDirectories()
    }

    const summaries: ClaudeSessionSummary[] = []
    for (const directory of directories) {
      let files: string[]
      try {
        files = readdirSync(directory)
      } catch {
        continue
      }
      for (const file of files) {
        if (!file.endsWith('.jsonl') || file.startsWith('.')) continue
        const summary = parseTranscript(join(directory, file))
        if (summary) summaries.push(summary)
      }
    }
    return summaries.sort((a, b) => b.modifiedAt.getTime() - a.modifiedAt.getTime()).slice(0, limit)
  }

  /**
   * The conversation so far. `claude --resume` under `--print` restores the
   * context for the model but prints nothing, so a resumed session would
   * arrive as an empty panel. The transcript on disk is the only copy.
   */
  transcript(sessionId: string, limit = 40): Turn[] {
    const path = this.transcriptPath(sessionId)
    if (!path) return []
    let contents: string
    try {
      contents = require('node:fs').readFileSync(path, 'utf8') as string
    } catch {
      return []
    }
    const turns: Turn[] = []
    for (const line of contents.split('\n')) {
      const object = parseLine(line)
      if (!object) continue
      const type = object.type
      if (type !== 'user' && type !== 'assistant') continue
      // Injected context, not something either party said.
      if (object.isMeta === true) continue
      const message = object.message as Json | undefined
      const text = extractText(message?.content)
      if (!text || !text.trim()) continue
      // Tool plumbing and harness scaffolding are not conversation.
      const lowered = text.toLowerCase()
      if (
        lowered.startsWith('<command') ||
        lowered.startsWith('caveat:') ||
        lowered.startsWith('<local-command') ||
        lowered.includes('</system-reminder>')
      ) {
        continue
      }
      const stamp = typeof object.timestamp === 'string' ? new Date(object.timestamp) : null
      turns.push({
        role: type === 'user' ? 'user' : 'assistant',
        text: text.slice(0, 4000),
        at: stamp && !Number.isNaN(stamp.getTime()) ? stamp : null,
      })
    }
    return turns.slice(-limit)
  }

  /** Transcripts are filed under a slug of the working directory, and a caller
   *  resuming by id does not necessarily know which one — so search. */
  private transcriptPath(sessionId: string): string | null {
    if (!/^[A-Za-z0-9-]+$/.test(sessionId)) return null
    for (const directory of this.projectDirectories()) {
      const candidate = join(directory, `${sessionId}.jsonl`)
      if (existsSync(candidate)) return candidate
    }
    return null
  }

  private projectDirectories(): string[] {
    try {
      return readdirSync(this.projectsDirectory, { withFileTypes: true })
        .filter((entry) => entry.isDirectory() && !entry.name.startsWith('.'))
        .map((entry) => join(this.projectsDirectory, entry.name))
    } catch {
      return []
    }
  }
}

function parseLine(line: string): Json | null {
  if (!line) return null
  try {
    const value = JSON.parse(line)
    return value && typeof value === 'object' ? (value as Json) : null
  } catch {
    return null
  }
}

/**
 * Pulls just enough out of a transcript to render a list row: the first user
 * message as a title, plus cwd and branch. Transcripts can be megabytes, so
 * this reads the head for metadata and counts lines without parsing the rest.
 */
function parseTranscript(path: string): ClaudeSessionSummary | null {
  let modifiedAt: Date
  try {
    modifiedAt = statSync(path).mtime
  } catch {
    return null
  }
  const head = readHead(path, 256 * 1024)
  if (!head.length) return null

  let cwd = ''
  let gitBranch: string | null = null
  let title: string | null = null
  for (const line of head.split('\n')) {
    const object = parseLine(line)
    if (!object) continue
    if (!cwd && typeof object.cwd === 'string') cwd = object.cwd
    if (gitBranch === null && typeof object.gitBranch === 'string' && object.gitBranch) gitBranch = object.gitBranch
    if (title === null && object.type === 'user') {
      const message = object.message as Json | undefined
      const candidate = extractText(message?.content)
      const cleaned = candidate ? summaryText(candidate) : null
      if (cleaned) title = cleaned
    }
    if (title !== null && cwd && gitBranch !== null) break
  }

  const id = basename(path, '.jsonl')
  const summary = (title || 'Untitled session').replace(/\n/g, ' ').trim().slice(0, 120)
  return {
    id,
    summary,
    cwd: cwd || basename(join(path, '..')),
    gitBranch,
    modifiedAt,
    messageCount: countLines(path),
  }
}

function readHead(path: string, bytes: number): string {
  let fd: number
  try {
    fd = openSync(path, 'r')
  } catch {
    return ''
  }
  try {
    const buffer = Buffer.alloc(bytes)
    const read = readSync(fd, buffer, 0, bytes, 0)
    return buffer.subarray(0, read).toString('utf8')
  } finally {
    closeSync(fd)
  }
}

function countLines(path: string): number {
  let fd: number
  try {
    fd = openSync(path, 'r')
  } catch {
    return 0
  }
  try {
    const buffer = Buffer.alloc(1 << 20)
    let total = 0
    for (;;) {
      const read = readSync(fd, buffer, 0, buffer.length, null)
      if (read <= 0) break
      for (let index = 0; index < read; index += 1) if (buffer[index] === 0x0a) total += 1
    }
    return total
  } finally {
    closeSync(fd)
  }
}

/** Transcripts often open with tooling scaffolding — command wrappers, caveat
 *  banners, injected reminders. Strip those and take the first line that reads
 *  like something a person typed. */
function summaryText(raw: string): string | null {
  let text = raw
  for (;;) {
    const open = text.indexOf('<')
    if (open < 0) break
    const close = text.indexOf('>', open + 1)
    if (close < 0) break
    const tag = text.slice(open + 1, close).split(' ')[0] ?? ''
    if (!tag || tag.startsWith('/')) break
    const end = text.indexOf(`</${tag}>`)
    text = end >= 0 ? text.slice(0, open) + text.slice(end + tag.length + 3) : text.slice(0, open) + text.slice(close + 1)
  }
  const ignored = ['caveat:', 'system-reminder', '<command', 'local-command']
  for (const line of text.split('\n')) {
    const trimmed = line.trim()
    if (trimmed.length <= 3) continue
    const lowered = trimmed.toLowerCase()
    if (ignored.some((prefix) => lowered.startsWith(prefix))) continue
    return trimmed
  }
  return null
}

/** Message content is either a plain string or an array of typed blocks. */
function extractText(content: unknown): string | null {
  if (typeof content === 'string') return content
  if (!Array.isArray(content)) return null
  for (const block of content) {
    if (block && typeof block === 'object' && (block as Json).type === 'text') {
      const text = (block as Json).text
      if (typeof text === 'string' && text) return text
    }
  }
  return null
}
